### PR TITLE
[WIP] Non-intrusive CMIS integration

### PR DESCRIPTION
**Pull request by joeribekker**
This work is done for Utrecht with a different approach than the current CMIS-integration which requires heavy refactoring on every API update.

### CODE REVIEW

File: src/openzaak/components/documenten/managers.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        django_date = datetime.datetime.fromtimestamp(timestamp).strftime('%Y-%m-%d')
+        return django_date
+
+
+def cmis_doc_to_django_model(cmis_doc):
+    from .models import EnkelvoudigInformatieObject, EnkelvoudigInformatieObjectCanonical
+
+    # The canonical field cannot be NULL
+    canonical = EnkelvoudigInformatieObjectCanonical.objects.create()
+
```
Comment (joeribekker) [inline src/openzaak/components/documenten/managers.py]: Also, on the `canonical` object we should set `lock` if needed.

`canonical.lock = cmis_doc.locked  # or lock needs to be retrieved seperately from Alfresco?`

File: src/openzaak/components/documenten/managers.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        timestamp = int(str(json_date)[:10])
+        django_date = datetime.datetime.fromtimestamp(timestamp).strftime('%Y-%m-%d')
+        return django_date
+
+
+def cmis_doc_to_django_model(cmis_doc):
+    from .models import EnkelvoudigInformatieObject, EnkelvoudigInformatieObjectCanonical
+
+    # The canonical field cannot be NULL
+    canonical = EnkelvoudigInformatieObjectCanonical.objects.create()
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openzaak/components/documenten/managers.py]: Modified to be `canonical = EnkelvoudigInformatieObjectCanonical()`

File: src/openzaak/components/documenten/managers.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+    date_fields = ['creatiedatum', 'ontvangstdatum', 'verzenddatum']
+    for date_field in date_fields:
+        date_value = getattr(cmis_doc, date_field)
+        if isinstance(date_value, int):
+            setattr(cmis_doc, date_field, convert_timestamp_to_django_date(date_value))
+
+    # Ensuring the charfields are not null
+    for field in EnkelvoudigInformatieObject._meta.get_fields():
+        if isinstance(field, fields.CharField) or isinstance(field, fields.TextField):
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openzaak/components/documenten/managers.py]: Added the `elif` to the loop:
```
    for field in EnkelvoudigInformatieObject._meta.get_fields():
        if isinstance(field, fields.CharField) or isinstance(field, fields.TextField):
            if getattr(cmis_doc, field.name) is None:
                setattr(cmis_doc, field.name, '')
        elif isinstance(field, fields.DateField):
            date_value = getattr(cmis_doc, field.name)
            if isinstance(date_value, int):
                setattr(cmis_doc, field.name, convert_timestamp_to_django_date(date_value))

```

File: src/openzaak/components/documenten/managers.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            "get() returned more than one %s -- it returned %s!" %
+            (self.model._meta.object_name, num)
+        )
+
+    def delete(self):
+        # Updating all the documents from Alfresco to have 'verwijderd=True'
+        number_alfresco_updates = 0
+        for django_doc in self._result_cache:
+            try:
+                self.get_cmis_client.delete_cmis_document(django_doc.uuid)
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openzaak/components/documenten/managers.py]: Went for the latter option (to use `settings.CMIS_DELETE_IS_OBLITERATE`)

File: src/openzaak/components/documenten/models.py [hunk 5]
```diff
@@ -232,6 +237,14 @@ def latest_version(self):
         versies = self.enkelvoudiginformatieobject_set.all()
         return versies.first()
 
+    def save(self, force_insert=False, force_update=False, using=None,
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openzaak/components/documenten/models.py]: Addressed this with by adding `lock_document()` and `unlock_document()` functions to `EnkelvoudigInformatieObjectCanonical`. These are called from the serializer when a document needs to be locked/unlocked. This is a work around for the fact that one cannot retrieve the uuid of a document from the canonical object when Alfresco is used.

File: src/openzaak/components/documenten/managers.py (hunk 1 above)
Comment (SilviaAmAm) [reply src/openzaak/components/documenten/managers.py]: Added canonical.lock = cmis_doc.versionSeriesCheckedOutId or ""

File: src/openzaak/components/documenten/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (joeribekker) [reply src/openzaak/components/documenten/models.py]: Well, workaround. I would say it's the best solution for the problem of supporting an external entity.

File: src/openzaak/components/documenten/managers.py (hunk 1 above)
Comment (joeribekker) [reply src/openzaak/components/documenten/managers.py]: I think this works, will review in depth details later.

File: src/openzaak/components/documenten/managers.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        if isinstance(field, fields.CharField) or isinstance(field, fields.TextField):
+            if getattr(cmis_doc, field.name) is None:
+                setattr(cmis_doc, field.name, '')
+
+    # Setting up a local file with the content of the cmis document
+    content_file = File(cmis_doc.get_content_stream())
+    content_file.name = cmis_doc.name
+    data_in_file = ContentFile(cmis_doc.get_content_stream().read())
+    content_file.content = data_in_file
+    content_file.path = private_media_storage.save(f'{content_file.name}', data_in_file)
```
[the remark this answers is not in the document]
Comment (joeribekker) [reply src/openzaak/components/documenten/managers.py]:
```python
# Inspiration from:
# - https://django-privates.readthedocs.io/en/latest/quickstart.html
# - https://github.com/jschneier/django-storages/
# - https://github.com/divio/django-storage-url/blob/master/django_storage_url/

# utils.py
from privates import 
from django.core.files.storage import Storage

# Similar to: https://github.com/jschneier/django-storages/blob/master/storages/backends/sftpstorage.py
class CMISStorageFile(File):
    # For now we can store file in memory via BytesIO, and later proxy it.
    # By doing this, sendfile also works so we don't need to change the viewset/download resource.

class CMISStorage(Storage):
    def _open(self, name, mode="rb"):
        return CMISStorageFile(name, mode, self)

    def url(self, name):
        # return url to file


class PrivateMediaStorageWithCMIS(LazyObject):

    def _setup(self):
        if settings.CMIS_ENABLED:
            self._wrapped = CMISStorage()
        else:
            self._wrapped = PrivateMediaFileSystemStorage()

private_media_storage_cmis = PrivateMediaStorageWithCMIS()


# models.py
from .utils import private_media_storage_cmis

class EnkelvoudigInformatieObject(AuditTrailMixin, APIMixin, InformatieObject):
    # added storage=...
    inhoud = PrivateMediaFileField(upload_to="uploads/%Y/%m/", storage=private_media_storage_cmis)
```

**Comment by joeribekker** [comment]
Replaced by #540

**Pull request closed without merging**